### ISSUE TITLE
Narrowing ??= assignment doesn't work when RHS contains ??

### ISSUE THREAD

**Issue by tomtheisen** [microsoft/TypeScript#44732]
# Bug Report

### 🔎 Search Terms

Control-Flow-Analysis Coalescing

### 🕗 Version & Regression Information

- The compiler is reporting a type error incorrectly.
- This changed between versions 3.9.7 and 4.0.5.
- Prior to 4.0, the syntax `??=` wasn't supported.
- It is unchanged through 4.4.0 nightly.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=4&ts=4.3.4#code/GYVwdgxgLglg9mABMAjACgIYH4BciDeAvgJSJ5EECwAUInYhAgM5SJIC8bIANt4llgKEA3DXqIM-LJzCja9AE4BTKCAVIMwxAHptbOIgAOCuACNuSgLY1CNGqEiwEyAEyZcQ0uUJV5dSQKcaGA8fAKecuLKquoSWrqIACoAnoZKiADkFAA+iOAAJkrAMGBK+RmIMEz6rBhMTDAA5mAY5ulQBlCp6VmEGQB0NnbUDtDwSDCFYLBdAGLgYwgAPIkAfJh4icSbvlEqahpytiMLTkjAAMzu3l5Cu-QB0pVTM8nzjuPBoVIRYor7sU0Oj0GG4TAMxVKQ2owyAA)

### 💻 Code

```ts
function f1(a?: {}) : {} {
    const n = null ?? {};
    a ??= n;
    return a; // no problem
}

function f2(a?: {}) : {} {
    a ??= (null ?? {});
    return a; // Type '{} | undefined' is not assignable to type '{}'.
}

function identityFunction<T>(a: T): T {
    return a;
}
function f3(a?: {}) : {} {
    a ??= identityFunction(null ?? {});
    return a; // also fine
}
```

### 🙁 Actual behavior

On line 9, a type error is reported: "Type '{} | undefined' is not assignable to type '{}'."

### 🙂 Expected behavior

I would expect that the code should pass type checking.  Here's why.

 1 . `??=` generally supports narrowing types, as shown in the playground link.
 2. `??` generally supports nullability narrowing, as shown in the playground link.
 3. There's nothing special about a RHS of a `??=` that contains a `??`.

**Comment by danielmayorga** [comment]
I believe this is also applicable to the logical OR assignment `||=`. The code in the playground mirrors the code and conditions stated above.

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=4&ts=4.3.4&ssl=21&ssc=1&pln=4&pc=1#code/GYVwdgxgLglg9mABMAjACgIYH4BciDeAvgJSJ5EECwAUInYhAgM5SJIC8bIANt4gD78ChANw16iDAP6cwY2vQBOAUyghFSDCMQB6HWziIADorgAjbsoC2NQjRqhIsBMgBMmXMNLlCVBXSlBTjQwHj5BL3kJFTUNSW09RAAVAE8jZUQAcgohcAATZWAYMGU8zMQYJgNWDCYmGABzMAwLDKhDKDSM7MJMgDpbe2pHaHgkGAKwWE6AMXBRhAAeJIA+TDwk4g2-aNV1TXk7YfnnJGAAZg8fb2Ed+kCZCsnplLmnMZCw6UjxJT24rS6fQYbhMQxFEqDahDIA)

I also believe there are situations where logical AND assignment `&&=` is inconsistent as well.

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=4&ts=4.3.4#code/GYVwdgxgLglg9mABMAjACgIYC5ECM5wA2AphmAJSI7AaEDOxiA3gLABQiniECdUiSALyIoAJxCMAZJOS0GAbnZdEGRNOFhFHLqOJQQopBnmIA9KYFxEAB1FxcJALbsAvu3ahIsBMgBMmHHwiUgoqWXpGVm1OVXVENDEJNRkaCPItZV19QxUTc0QAFQBPa0YAciCSMjLEGDpLfgw6OhgAczAMB0YoKygS8tSGMoA6V3c2T2h4JBgAE2IwWD6AMXAphAAeAoA+AMLyHALmJR09AyMtNzYPNe8kYABmPcqQymo5SJOY5OE5haWiqsvNM0IMpDJEsR0l9EFlzrkzBY5FZgDAwMQxtcsUA)

I'm trying to look through the codebase to learn about it and how to fix this issue, and stumbled across these cases as well. Don't want to open a new work item since they are so closely related. Let me know if there is a better procedure to report this.

**Comment by fatcerberus** [comment]
> I believe this is also applicable to the logical OR assignment ||=. The code in the playground mirrors the code and conditions stated above.

If so, this is probably related to #44553.

**Comment by AlCalzone** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/41153 and possibly https://github.com/microsoft/TypeScript/issues/40494 ?

**Comment by tomtheisen** [comment]
Seems like this is fixed, as of 5.6.2
